File: Quinto Semestre ESCOM/Algoritmos/Códigos/subsetSumProblem/DP/subsetSumDP.py
def esSumaConjunto(conjunto, n, sum):
    #*Se crea la tabla inicializando los valores subconjunto[n+1][sum+1]
    subconjunto =([[False for i in range(sum + 1)]
            for i in range(n + 1)])
    
    #*Si la suma a buscar es 0, se regresa true
    for i in range(n + 1):
        subconjunto[i][0] = True

    #*Si la suma no es 0, pero el conjunto es vacio se regresa false
    for i in range(1, sum + 1):
        subconjunto[0][i]= False
            
    #*Se llena la tabla recorriendo la suma y los elementos del conjunto
    for i in range(1, n + 1):
        for j in range(1, sum + 1):
            #*Si el elemento actual es mayor al valor actual de la suma copiamos el valor del caso anterior
            if j<conjunto[i-1]:
                subconjunto[i][j] = subconjunto[i-1][j]
            #*Si el valor actual de la suma es mayor que el iesimo elemento
            #*hacemos or entre los casos previos y el caso de j-valor de suma actual
            if j>= conjunto[i-1]:
                subconjunto[i][j] = (subconjunto[i-1][j] or
                                subconjunto[i - 1][j-conjunto[i-1]])

    return subconjunto[n][sum]

conjunto = [2,6,4,8]
sum = 10
n = len(conjunto)
if (esSumaConjunto(conjunto, n, sum) == True):
    print("Se encontro un subconjunto con la suma dada")
else:
    print("Se encontro un subconjunto con la suma dado")


tab = [[-1 for i in range(2000)] for j in range(2000)]

def esSumaConjunto(conjunto, n, sum):

    if (sum == 0):
        return 1
    
    if (n <= 0):
        return 0

    if (tab[n - 1][sum] != -1):
        return tab[n - 1][sum]

    if (conjunto[n - 1] > sum):
        tab[n - 1][sum] = esSumaConjunto(conjunto, n - 1, sum)
        return tab[n - 1][sum]
    else:
        tab[n - 1][sum] = esSumaConjunto(conjunto, n - 1, sum)
        return tab[n - 1][sum] or esSumaConjunto(conjunto, n - 1, sum - conjunto[n - 1])

conjunto = [2,6,4,8]
sum = 10
n = len(conjunto)
if (esSumaConjunto(conjunto, n, sum) == True):
    print("Se encontro un subconjunto con la suma dada")
else:
    print("Se encontro un subconjunto con la suma dado")

File: Quinto Semestre ESCOM/Algoritmos/Códigos/subsetSumProblem/DP/subsetSumDp.java
class SSP {

    static boolean esSumaConjunto(int conjunto[], int n, int sum)
    {
        boolean subconjunto[][] = new boolean[n + 1][sum + 1];

        //*Si la suma a buscar es 0, se regresa true
        for(let i=0; i<= n; i++)
        subconjunto[i][0] = true;

        //*Si la suma no es 0, pero el conjunto es vacio se regresa false
        for(let i = 1; i <= sum; i++)
            subconjunto[0][i] = false;

        //*Se llena la tabla recorriendo la suma y los elementos del conjunto
        for(let i=1; i <= n; i++){
            for(let j=1; j <= sum; j++){
                //*Si el elemento actual es mayor al valor actual de la suma copiamos el valor del caso anterior
                if(j < conjunto[i - 1])
                    subconjunto[i][j] = subconjunto[i-1][j];
                //*Si el valor actual de la suma es mayor que el iesimo elemento
                //*]hacemos or entre los casos previos y el caso de j-valor de suma actual
                if(j >= conjunto[i-1])
                    subconjunto[i][j] = subconjunto[i-1][j] || subconjunto[i-1][j-conjunto[i-1]];
            }
        }

        return subconjunto[n][sum];
    }

    public static void main(String args[])
    {
        int conjunto[] = { 2,6,4,8 };
        int sum = 10;
        int n = conjunto.length;
        if (esSumaConjunto(conjunto, n, sum) == true)
            System.out.println("Se encontro un subconjunto con la suma dada");
        else
            System.out.println("Se encontro un subconjunto con la suma dado");
    }
}


class SSP {

    static boolean esSumaConjunto(int conjunto[], int n, int sum)
    {
        int tab[][] = new int[n + 1][sum + 1];
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= sum; j++) {
                tab[i][j] = -1;
            }
        }

        if (sum == 0)
            return 1;

        if (n <= 0)
            return 0;

        if (tab[n - 1][sum] != -1)
            return tab[n - 1][sum];

        if (conjunto[n - 1] > sum)
            return tab[n - 1][sum]
                = esSumaConjunto(conjunto, n - 1, sum);
        else {
            if (esSumaConjunto(conjunto, n - 1, sum) != 0 || esSumaConjunto(conjunto, n - 1, sum - conjunto[n - 1]) != 0) {
                return tab[n - 1][sum] = 1;
            }
            else
                return tab[n - 1][sum] = 0;
        }
    }

    public static void main(String args[])
    {
        int conjunto[] = { 2,6,4,8 };
        int sum = 10;
        int n = conjunto.length;
        if (esSumaConjunto(conjunto, n, sum) == true)
            System.out.println("Se encontro un subconjunto con la suma dada");
        else
            System.out.println("Se encontro un subconjunto con la suma dado");
    }
}